halo2_proofs/src/dev.rs

```diff
@@ -395,25 +421,29 @@ impl<F: Field + Group> Assignment<F> for MockProver<F> {
         A: FnOnce() -> AR,
         AR: Into<String>,
     {
-        if !self.usable_rows.contains(&row) {
-            return Err(Error::not_enough_rows_available(self.k));
-        }
+        if self.in_phase(FirstPhase) {
+            if !self.usable_rows.contains(&row) {
+                return Err(Error::not_enough_rows_available(self.k));
+            }
 
-        if let Some(region) = self.current_region.as_mut() {
-            region.update_extent(column.into(), row);
-            region
-                .cells
-                .entry((column.into(), row))
-                .and_modify(|count| *count += 1)
-                .or_default();
+            if let Some(region) = self.current_region.as_mut() {
+                region.update_extent(column.into(), row);
+                region
+                    .cells
+                    .entry((column.into(), row))
+                    .and_modify(|count| *count += 1)
+                    .or_default();
+            }
         }
 
-        *self
-            .advice
-            .get_mut(column.index())
-            .and_then(|v| v.get_mut(row))
-            .ok_or(Error::BoundsFailure)? =
-            CellValue::Assigned(to().into_field().evaluate().assign()?);
+        if self.in_phase(column.column_type().phase) {
+            *self
+                .advice
+                .get_mut(column.index())
+                .and_then(|v| v.get_mut(row))
+                .ok_or(Error::BoundsFailure)? =
+                CellValue::Assigned(to().into_field().evaluate().assign()?);
+        }
```

Comment:
What if we aren't in the correct Phase? Should we log something or similar? Or panic in case the behaviour is unexpected?

Reply:
In wrong phase there could be 2 kind of case:

1. Assign to previous phase
    I think this is legal in most cases (since in circuit we don't check if the phase is correct or not, we always assign it). 
2. Assign to later phase
    I think this is also legal in some cases, for example we could assign a constant to phase 2 column, which will always have value.

So instead what might be doable is to cache previous phase assigned values and make sure they are always the same, but it might take more time to collect all the values.

I'll try this and see if it works well.

Reply:
We can try in a follow-up PR so that this is not blocked! 
I just thought it could make sense for us to test it!

Reply:
Added the check in https://github.com/privacy-scaling-explorations/halo2/pull/129/commits/a5b42f7087982387082d3fc59b09ec1ee7b80442.